

---
title: Fig5. Peripheral immune cells atlas and inflammatory-related mechanisms in COVID-19
  revealed by CLEAR.
author: "Yuqi Cheng"
output: html_document
fontsize: 24pt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Self-supervised Contrastive learning for integrative single cell scRNA-seq data analysis. Fig 5 Rmarkdown.
Please see our publication at: xxx

This Rmarkdown file help to walk through the reproduction process of figure 5.

## Load Package

```{r lib, warning=FALSE, error=FALSE, message=FALSE, eval=FALSE}
library(SingleCellExperiment)
library(scater)
library(scran)
library(pheatmap)
library(Seurat)
library(magrittr)
library(RColorBrewer)
library(ggsci)
library(SeuratDisk)
library(gridExtra)
library(loomR)
library(ggpubr)
library(ggforce)
library(org.Hs.eg.db)
library(DOSE)
library(enrichplot)
library(clusterProfiler)
```

## Load dataset and annotate cell type

We use the COVID-19 dataset processed by CLEAR

```{r load_data, eval=FALSE}
memory.limit(1e+10)
sce <- readRDS('~/R Scripts/rna_clustering/covid/pbmc_covid.rds')

sce <- FindNeighbors(sce, dims = 1:128, reduction = 'scarlet')
sce <- FindClusters(sce, resolution = 1.2)
sce <- RunUMAP(sce, dims = 1:128, reduction = 'scarlet')

#find markers
all.marker <- FindAllMarkers(sce, only.pos = TRUE, )
write.csv(all.marker, 'marker_genes.csv', row.names = F)

#Simplify
new_cluster.ident <- c('CD14 Monocyte','B cell','CD4 T cell','CD4 T cell','NK','CD8 T cell',
                       'CD8 T cell','CD8 T cell','NK','CD14 Monocyte','CD14 Monocyte','CD14 Monocyte',
                       'Plasmablast','CD4 T cell','Plasmablast','CD16 Monocyte','CD4 T cell',
                       'RBC','CD8 T cell','Plasmablast','Platelet','CD8 T cell',
                       'NK','CD16 Monocyte','Neutrophil','B cell','DC','RBC',
                       'SC & Eosinophil','pDC','Neutrophil','CD4 T cell')

names(new_cluster.ident) <- levels(sce)
sce <- RenameIdents(sce, new_cluster.ident)
```

## Generate a and b

```{r fig5ab, eval=FALSE}
pal <- c("#A6CEE3","#4393C3","#B2DF8A","#41AB5D","#FB9A99","#D6604D","#FDBF6F",
         "#FF7F00","#CAB2D6","#6A3D9A","#FFFF99","#B15928","#737373")

custom_fill_colors = c(RColorBrewer::brewer.pal(9, "Blues")[7], 
                       RColorBrewer::brewer.pal(9, "Oranges")[3],
                       RColorBrewer::brewer.pal(9, "Reds")[7])

DimPlot(sce, label = T, label.size = 4, cols = pal) +
  labs(x = "UMAP1", y = "UMAP2") + 
  theme(legend.text = element_text(size = 13),
        legend.title = element_text(size = 11)) +
  theme(axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank())

DimPlot(sce, group.by = 'Ventilated', cols = custom_fill_colors) + 
  labs(x = "UMAP1", y = "UMAP2") + 
  theme(legend.text = element_text(size = 13),
        legend.title = element_text(size = 11)) +
  theme(axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank())
```

## Generate c

We mainly use 

```{r fig5c, eval=FALSE}
cell_type_health <- Idents(subset(sce, Ventilated == 'Healthy')) %>% table()
cell_type_Nonven <- Idents(subset(sce, Ventilated == 'NonVent')) %>% table()
cell_type_Vent <- Idents(subset(sce, Ventilated == 'Vent')) %>% table()

total <- data.frame(row.names = cell_type$Var1, 
           'Heatlhy' = as.numeric(cell_type_health/cell_type$Freq), 
           'NonVent' = as.numeric(cell_type_Nonven/cell_type$Freq), 
           'Ventilated' = as.numeric(cell_type_Vent/cell_type$Freq))

#write.csv(total, 'total.csv')

#total <- read.csv('~/R Scripts/plotting/total.csv')
#read supplementary table 2

p2 <- ggplot(total, aes(x = Cell_Type, y = Percentage, fill = Status)) +
  geom_col(position = 'fill', width = 0.8) +
  xlab("Cell Type") + ylab("Percentage") +
  scale_fill_manual(values = custom_fill_colors) +theme_bw()+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_blank(), axis.line = element_line(colour = "black", size = 0.6))+
  theme(axis.title.x = element_blank(),
        axis.title.y = element_text(size = 14),
        axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
        axis.text.y = element_text(size = 12)) +
  theme(legend.text = element_text(size = 11),
        legend.title = element_text(size = 11),
        legend.position = 'top') +
  scale_y_continuous(expand=c(0,0)) + 
  scale_x_discrete(labels = c('B cell' = 'B cell (3548)','CD14 Monocyte' = 'CD14 Monocyte (9184)',
                              'CD16 Monocyte' = 'CD16 Monocyte (1694)', 'CD4 T cell' = 'CD4+ T cell (7986)',
                              'CD8 T cell' = 'CD8+ T cell (8814)', 'DC' = 'DC (351)','Neutrophil' = 'Neutrophil (619)',
                              'NK' = 'NK (5295)','pDC' = 'pDC (243)', 'Plasmablast' = 'Plasmablast (3910)', 'RBC' = 'RBC (1075)',
                              'Platelet' = 'Platelet (715)','SC & Eosinophil' = 'SC & Eosinophil (261)'))

p2
```

## Genrate d

```{r fig5d,eval=FALSE}
#vlnplot
feature <- c('CD3E','CD4','CD8A','IL7R',
             'PRF1','NKG7','NCAM1',
             'CD14','LYZ','FCGR3A','MS4A7',
             'MS4A1','IGHM','IGHG1','IGHG3',
             'HBB','HBA1',
             'PPBP','TUBB1',
             'NAMPTL','NAMPT',
             'FCER1A',
             'IRF8', 'TCF4',
             'GATA2', 'MYB')

sce@active.ident <- factor(x = sce@active.ident, 
                           levels = c("CD4 T cell", "CD8 T cell","NK", "CD14 Monocyte", "CD16 Monocyte", 
                                      "B cell","Plasmablast", "RBC","Platelet","Neutrophil",
                                      "DC","SC & Eosinophil","pDC"))

pal_vln <- c("#B2DF8A","#FB9A99","#41AB5D","#A6CEE3","#FDBF6F","#4393C3","#D6604D",
         "#FF7F00","#CAB2D6","#6A3D9A","#FFFF99","#B15928","#737373")

VlnPlot(sce,features = feature, pt.size = 0, stack = T, flip = F,  fill.by = 'ident', cols = pal_vln)+ NoLegend()+
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_blank())
```

## Generate e

```{r fig5e, eval=FALSE}
pal <- c("#A6CEE3","#4393C3","#B2DF8A","#41AB5D","#FB9A99","#D6604D","#FDBF6F",
         "#FF7F00","#CAB2D6","#6A3D9A","#FFFF99","#B15928","#737373")

pal_proin <- c('lightgrey','lightgrey','lightgrey',"#A6CEE3","#FDBF6F",'lightgrey','lightgrey',
         'lightgrey','lightgrey','lightgrey','lightgrey','lightgrey','lightgrey')

## Generate main panel
DimPlot(sce, label = F, label.size = 4, cols = pal_proin,pt.size = 0.01) +labs(x = "UMAP1", y = "UMAP2") + 
  theme_bw() +
  theme(legend.text = element_text(size = 13),
        legend.title = element_text(size = 11)) +
  theme(axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(), 
        axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(),
        panel.grid = element_blank())

## Generate cytolines panels
mono <- subset(sce,idents = c('CD16 Monocyte','CD14 Monocyte'))

FeaturePlot(mono, 'TNF', pt.size = 0.1)
FeaturePlot(mono, 'IL10', pt.size = 0.1)
FeaturePlot(mono, 'IL1B', pt.size = 0.1)
FeaturePlot(mono, 'IL6', pt.size = 0.1)
```

## Generate f

We use wilcox method to perform DEG analysis.

```{r fig5f, eval=FALSE}
#CD14+ monocyte DEG
mon_health <- subset(sce, idents = 'CD14 Monocyte')
mon_deg <- FindMarkers(mon_health, ident.1 = 'NonVent', ident.2 = 'Healthy', logfc.threshold = 0.1,  
                       group.by = 'Ventilated', only.pos = F, test.use = "wilcox")
mon_deg_vent <- FindMarkers(sce, ident.1 = 'Vent', ident.2 = 'Healthy', group.by = 'Ventilated', logfc.threshold = 0.1
                            subset.ident = 'CD14 Monocyte',only.pos = F, test.use = 'wilcox')

mon_deg <- data.frame(NonVent = mon_deg$avg_log2FC, row.names = rownames(mon_deg))
mon_deg$Gene1 <- rownames(mon_deg)
mon_deg_vent <- data.frame(Vent = mon_deg_vent$avg_log2FC, row.names = rownames(mon_deg_vent))
mon_deg_vent$Gene2 <- rownames(mon_deg_vent)

intersect_gene <- intersect(rownames(mon_deg), rownames(mon_deg_vent))
mon_deg[intersect_gene,]

total <- data.frame(NonVent = mon_deg[intersect_gene,]$NonVent, 
                    Vent = mon_deg_vent[intersect_gene,]$Vent, 
                    row.names = intersect_gene)

heatmap_genes <- read.csv('~/R Scripts/rna_clustering/covid/heatmap.csv')
heatmap_genes$Gene.to.check
total <- total[heatmap_genes$Gene.to.check,]
total <- na.omit(total)

total[order(total$NonVent, decreasing = F),]

total_cd14 <- total

bk <- c(seq(-6,-0.1,by=0.01),seq(0,6,by=0.01))
annotation_col <- data.frame(Treatment = c('Moderate','Severe (ARDS)'), row.names = c('NonVent','Vent'))
pheatmap(total_cd14, 
         color = c(colorRampPalette(colors = c("#2171B5","white"))(length(bk)/2),
                   colorRampPalette(colors = c("white","#CB181D"))(length(bk)/2)),
         scale = 'column',cluster_cols=F,cluster_rows = T,
         treeheight_col = 15,
         legend_breaks = seq(-5,5,1),
         show_rownames = T,
         annotation_col = annotation_col,
         show_colnames = F,
         cutree_rows = 2,
         gaps_row = 10,
         annotation_legend = T,
         border = F)
```

## Generate fig5 g h

```{r fig5gh, eval=FALSE}
# ISG score
isg.list <- read.csv('~/R Scripts/rna_clustering/covid/isg.csv')

sce <- AddModuleScore(sce, features = list(as.character(isg.list$ISG.Sig)), name = 'ISG1')
sce_deg$ISG1

sce_cd14 <- subset(sce, idents = 'CD14 Monocyte')
sce_cd16 <- subset(sce, idents = 'CD16 Monocyte')

#cd 14 mono isg score
## health
h <- subset(sce_cd14, Donor_full == 'H6')
sum(h$IFN1)/dim(h)[2]

##moderate
c1a <- subset(sce_cd14, Donor_full == 'C3')
sum(c1a$IFN1)/dim(c1a)[2]

##severe
c1b <- subset(sce_cd14, Donor_full == 'C7')
sum(c1b$IFN1)/dim(c1b)[2]

#cd 16 mono isg score
## health
h <- subset(sce_cd16, Donor_full == 'H6')
sum(h$IFN1)/dim(h)[2]

##moderate
c1a <- subset(sce_cd16, Donor_full == 'C6')
sum(c1a$IFN1)/dim(c1a)[2]

##severe
c1b <- subset(sce_cd16, Donor_full == 'C7')
sum(c1b$IFN1)/dim(c1b)[2]

#plot cd 14 mono isg score
isg.score <- read.csv('~/R Scripts/rna_clustering/covid/isg.score.cd14.csv')

ggboxplot(isg.score, x = 'Status', y = 'ISG.Score', fill = 'Status', 
          palette = custom_fill_colors, bxp.errorbar = T, width = 0.5, bxp.errorbar.width = 0.3) +
  geom_signif(comparisons = list(c("HD", "Moderate"),
                                 c("Moderate", "Severe (ARDS)"),
                                 c("HD", "Severe (ARDS)")),
              y_position = c(0.28,0.29,0.30),
              map_signif_level=F,
              margin_top = 0.01)+
  theme_bw() +
  theme(axis.text.x=element_text(size = 12),
        axis.text.y=element_text(size = 12),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlab(NULL)

#################################

#HLA score
memory.limit(1e11)
hla.list <- read.csv('~/R Scripts/rna_clustering/covid/hla.list.csv')
sce <- AddModuleScore(sce, features = list(as.character(hla.list$HLA.genes)), name = 'HLA_all')
sce_cd14 <- subset(sce, idents = 'CD14 Monocyte')
sce_cd16 <- subset(sce, idents = 'CD16 Monocyte')

#cd 14 mono hla score
## health
h <- subset(sce_cd14, Donor_full == 'H6')
sum(h$HLA_all1)/dim(h)[2]

## moderate
c1a <- subset(sce_cd14, Donor_full == 'C6')
sum(c1a$HLA_all1)/dim(c1a)[2]

## severe
c1b <- subset(sce_cd14, Donor_full == 'C1 B')
sum(c1b$HLA_all1)/dim(c1b)[2]

#cd 16 mono hla score
## health
h <- subset(sce_cd16, Donor_full == 'H6')
sum(h$HLA_all1)/dim(h)[2]

## moderate
c1a <- subset(sce_cd16, Donor_full == 'C6')
sum(c1a$HLA_all1)/dim(c1a)[2]

## severe
c1b <- subset(sce_cd16, Donor_full == 'C7')
sum(c1b$HLA_all1)/dim(c1b)[2]

#plotting
hla.score <- read.csv('~/R Scripts/rna_clustering/covid/hla.score.cd14.csv')

ggboxplot(hla.score, x = 'Status', y = 'HLA.Score', fill = 'Status', 
          palette = custom_fill_colors, bxp.errorbar = T, width = 0.5, bxp.errorbar.width = 0.3) +
  geom_signif(comparisons = list(c("HD", "Moderate"),
                                 #c("Moderate", "Severe (ARDS)"),
                                 c("HD", "Severe (ARDS)")),
              y_position = c(0.56,0.60),
              map_signif_level=F,
              margin_top = 0.01)+
  theme_bw() +
  theme(axis.text.x=element_text(size = 12),
        axis.text.y=element_text(size = 12),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlab(NULL) +
  scale_y_continuous(limits = c(0.0,0.64), breaks = seq(0.0,0.6,0.2))

```

## Generate i

```{r fig5i, eval=FALSE}
#gsra
geneList<-mon_deg_vent$Vent
names(geneList)=mon_deg_vent$Gene2
geneList <- sort(geneList,decreasing = T)

go <- read.gmt('~/R Scripts/rna_clustering/covid/c5.go.bp.v7.4.symbols.gmt')
cd14_m_h_go <- GSEA(geneList = geneList, TERM2GENE = go)
cd14_m_h_go@result

cd14_m_h_go@result[order(cd14_m_h_go@result$enrichmentScore,decreasing = T),]
write.csv(cd14_m_h_go@result,'go_mono14_sever.csv')
gseaplot2(cd14_m_h_go,c(2,19,27,23,24,3,4,5,6,1),pvalue_table = F, ES_geom = 'line',
          rel_heights = c(1.4, 0.5, 1),base_size = 12, title = 'Upregulated Pathway')
```

